File: zs-leetcode-go/SymmetricTree/SymmetricTree.go
package SymmetricTree

import "for-an-offer/zs-leetcode-go/types"

func isSymmetric(root *types.TreeNode) bool {
	if root == nil {
		return false
	}
	return auxFunc(root.Left, root.Right)
}

func auxFunc(node1, node2 *types.TreeNode) bool {
	if node1 == nil && node2 != nil || node1 != nil && node2 == nil {
		return false
	}
	if node1 == nil && node2 == nil {
		return true
	}
	if node1.Val == node2.Val {
		return auxFunc(node1.Left, node2.Right) && auxFunc(node1.Right, node2.Left)
	}
	return false
}

File: sjq-leetcode-C++/symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(root == NULL) return true;
        return isSame(root->left, root->right);
    }
    bool isSame(TreeNode* a, TreeNode*b){
        if(a == NULL) return !b;
        else if(b==NULL) return !a;
        else
            return (a->val==b->val) and isSame(a->left, b->right) and isSame(a->right, b->left);
    }
};
